lib/rules/no-unstable-nested-components.js

```diff
@@ -166,6 +166,30 @@ function isReturnStatementOfHook(node, context) {
   );
 }
 
+/**
+ * Check whether given node is declared inside a specific function.
+ *
+ * ```jsx
+ * {intl.formatMessage(messages.testMessage, {
+ *   a: (...chunks) => <a href="/">{chunks}</a>
+ *   // ^ This returns JSX, so technically it's a component
+ * })}
+ * ```
+ * @param {ASTNode} node The AST node
+ * @param {Context} context eslint context
+ * @param {Array<string>} ignoredFunctions
+ * @returns {Boolean} True if component is declared inside a render prop, false if not
+ */
+function isDescendantOfIgnoredFunction(node, context, ignoredFunctions) {
+  const code = context.getSourceCode();
+  return context.getAncestors().some((n) => ignoredFunctions.some((functionName) => {
+    if (n.type === 'CallExpression') {
+      return code.getText(n.callee) === functionName;
+    }
```

Comment:
Is `code.getText(n.callee)` an ok solution here? Are there better ways to handle this check? I figure this could break if you format your code awkwardly. :man_shrugging: Since we're really working around a bad API that we can't change, maybe this is fine. What do you think?

```
intl
  .formatMessage()
```

Reply:
I don't see a cleaner way to get the whole `intl.formatMessage`. The code below works, but only for exactly `object.property()` type calls. This code can't handle handle `ignoreFunctions: [ 'myCustomFunction', 'why.would.anyone.do.this' ]`. Especially since my initial PR targeted exactly and exclusively `intl.formatMessage`, I'm trying to keep this a bit more flexible. 

```
      if (n.callee.type === 'MemberExpression') {
        if (n.callee.object.type === 'Identifier' && n.callee.property.type === 'Identifier') {
          return `${n.callee.object.name}.${n.callee.property.name}` === functionName;
        }
      }
```

![image](https://user-images.githubusercontent.com/324999/225117721-789671a8-7d24-4bd7-bc77-f1a56ef39633.png)

Considering this is all working around an API that we agree is bad and we can't change, do you think it's ok to stick with `getText`? 

A couple questions that come to mind for me that I don't know how to answer are: 
- Is it expensive to call `context.getSourceCode()`? 
- Should it only be called once somewhere else, or is it better to keep it with the feature that uses it?

Reply:
It's not expensive, and it's fine either way. I generally put it in a variable when it's used multiple times, but i don't think it makes a material difference.

The challenge with the getText approach is that it's very brittle - you'd need a bunch of other linting rules to ensure that a particular function is required/imported with the right name, destructured with the right name or not, etc, in order for any config value to be meaningful.

Reply:
I agree that you have a point about it being brittle, but I'm not sure how solid this really needs to be. 

For my codebase we simply have a convention of using `const intl = useIntl()` and `intl.formatMessage()`. I'm fine with this rule not catching `const { formatMessage } = useIntl()`. We use code owners on the lint rules, so if someone adds `formatMessage` to the rule, we'll request they change the source instead and follow the naming convention.

If this can be achieved through types, I'm really looking for target this function. `import('react-intl').IntlFormatters['formatMessage']`. Without types, it seems like a long journey to work through:  `the method formatMessage` on `the object returned` from `the useIntl() import` on ` react-intl module`. Are there existing tools within the repo that would do some parts of that?

Reply:
both eslint itself, and this plugin, have some utility functions to help you track that, but unfortunately nothing super trivial to use.

essentially you'd have to record all requires/imports and the identifier used, and then find all the potential call expressions and store those, and then at program exit, check if any of the candidates you found came from the specifier in question.

Reply:
If that's the hurdle that needs to be cleared here, I think I'll take a change in direction. I'd actually like a rule closer to `react-hooks/exhaustive-deps` that checks the stability of the JSX call.

```
export function Component() {
  const SomeComponent = () => <div />;
  //    ^ The 'SomeComponent' function makes the dependencies of useEffect Hook
  //    (at line 41) change on every render. To fix this, wrap the definition of
  //    'SomeComponent' in its own useCallback() Hook.
  //    - react-hooks/exhaustive-deps

  React.useEffect(() => {
    SomeComponent();
  }, [SomeComponent]);

  return (
    <SomeComponent />
  // ^ Idea: "SomeComponent" has an unstable identity, causing it to un-mount
  // and re-mount instead of re-render
  );
}
```

Reply:
I think we might have that already? https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unstable-nested-components.md

Reply:
Yes, that's this plugin. But I can't use it because our codebase uses `intl.formatMessage()` and there are too many false positives. It's also written into our coding standards, so the problem isn't going away.


The project I'm working on has an internal ESLint plugin. I have something there that examines each JSXIdentifier and tries to figure out if it's stable. It makes some assumptions, but it's working well enough for what I need. It assumes anything from a hook is stable, anything from an outer scope, or the current scope's parameters are stable. I assume anything inside a `describe/test` function doesn't need to be scanned.